services/src/main/java/org/exoplatform/webconferencing/WebConferencingService.java

```diff
@@ -0,0 +1,1235 @@
+                                   String ownerId,
+                                   String ownerType) {
+    // Synchronize on userListeners to have a consistent list of listeners to fire
+    Set<UserCallListener> listeners = userListeners.get(userId);
+    if (listeners != null) {
+      for (UserCallListener listener : listeners) {
+        // TODO we may lose events: when one request is completing pooling with some event, the listener will
+        // stop listen and next one will be registered after some period - this time events will not be
+        // delivered to the client.
+        // As a solution, we need a temporal pool to save (deferred) events for given user
+        // (listener.getUserId()) until it will send a new request or the pool expired
+        if (listener.isListening()) {
+          listener.onCallState(callId, providerType, callState, ownerId, ownerType);
+        }
+      }
+    }
+  }
+
+  /**
+   * Fire user call joined a new part.
+   *
+   * @param callId the call id
+   * @param providerType the provider type
+   * @param ownerId the owner id
+   * @param ownerType the owner type
+   * @param partId the part id
+   * @param userId the user id
+   */
+  protected void fireUserCallJoined(String callId,
+                                    String providerType,
+                                    String ownerId,
+                                    String ownerType,
+                                    String partId,
+                                    String userId) {
+    Set<UserCallListener> listeners = userListeners.get(userId);
+    if (listeners != null) {
+      for (UserCallListener listener : listeners) {
+        if (listener.isListening()) {
+          listener.onPartJoined(callId, providerType, ownerId, ownerType, partId);
+        }
+      }
+    }
+  }
+
+  /**
+   * Fire user call part leaved.
+   *
+   * @param callId the call id
+   * @param providerType the provider type
+   * @param ownerId the owner id
+   * @param ownerType the owner type
+   * @param partId the part id
+   * @param userId the user id
+   */
+  protected void fireUserCallLeaved(String callId,
+                                    String providerType,
+                                    String ownerId,
+                                    String ownerType,
+                                    String partId,
+                                    String userId) {
+    Set<UserCallListener> listeners = userListeners.get(userId);
+    if (listeners != null) {
+      for (UserCallListener listener : listeners) {
+        if (listener.isListening()) {
+          listener.onPartLeaved(callId, providerType, ownerId, ownerType, partId);
+        }
+      }
+    }
+  }
+
+  /**
+   * Adds the plugin.
+   *
+   * @param plugin the plugin
+   */
+  public void addPlugin(ComponentPlugin plugin) {
+    Class<CallProvider> pclass = CallProvider.class;
+    if (pclass.isAssignableFrom(plugin.getClass())) {
```

Comment:
Cannot the addPlugin methods directly accept CallProvider class instead of ComponentPlugin ?

Reply:
It can. But then, in case of wrong new plugin (configuration or class), it will fail on the server start. Do you want let a provider plugin let interrupt the server start? Currently server will continue to start but warn in the configuration - a safer approach.

Reply:
We could say the same thing if the class is not a ComponentPlugin, it will interrupt the server start :) But ok, I get your point, that's fine for me.